File: architecture-skills/design-patterns/examples/strategy/strategy.go
// Strategy pattern in Go.
//
// Go usually skips the one-method interface: a named function type is
// the strategy, plain functions and closures implement it (compare
// http.HandlerFunc). An interface is only needed when strategies have
// several methods.

package main

import "fmt"

type Order struct {
	WeightKg   float64
	DistanceKm float64
}

// ShippingStrategy is the strategy: any func with this signature fits.
type ShippingStrategy func(Order) float64

func flatRate(Order) float64 {
	return 5.00
}

func weightBased(o Order) float64 {
	return 4.00 + 1.50*o.WeightKg
}

func distanceBased(o Order) float64 {
	return 2.00 + 0.05*o.DistanceKm
}

// A stateful strategy: a closure captures the negotiated rates.
func carrierRate(base, perKg float64) ShippingStrategy {
	return func(o Order) float64 {
		return base + perKg*o.WeightKg
	}
}

// Checkout is the context: it delegates pricing to the current strategy.
type Checkout struct {
	strategy ShippingStrategy
}

func (c *Checkout) SetStrategy(s ShippingStrategy) {
	c.strategy = s
}

func (c *Checkout) ShippingFor(o Order) float64 {
	return c.strategy(o)
}

func main() {
	order := Order{WeightKg: 2.4, DistanceKm: 120}

	strategies := []struct {
		name     string
		strategy ShippingStrategy
	}{
		{"flat", flatRate},
		{"weight", weightBased},
		{"distance", distanceBased},
		{"carrier", carrierRate(3.00, 2.00)},
	}

	checkout := &Checkout{strategy: flatRate}
	for _, s := range strategies {
		checkout.SetStrategy(s.strategy)
		fmt.Printf("%8s: $%.2f\n", s.name, checkout.ShippingFor(order))
	}
}

File: architecture-skills/design-patterns/examples/strategy/strategy.swift
// Strategy pattern in Swift.
//
// Closures cover stateless strategies (sorted(by:) takes one), so the
// classic protocol form is only needed when a strategy carries state or
// several requirements. Both are shown; the struct plugs in by passing
// its method as a closure.

import Foundation

struct Order {
    let weightKg: Double
    let distanceKm: Double
}

// --- Closure form: idiomatic for stateless strategies ---

typealias ShippingStrategy = (Order) -> Double

let flatRate: ShippingStrategy = { _ in 5.00 }
let weightBased: ShippingStrategy = { 4.00 + 1.50 * $0.weightKg }
let distanceBased: ShippingStrategy = { 2.00 + 0.05 * $0.distanceKm }

// --- Protocol form: for stateful strategies ---

protocol ShippingCost {
    func cost(for order: Order) -> Double
}

/// A stateful strategy: the negotiated rates are its state.
struct CarrierRate: ShippingCost {
    let base: Double
    let perKg: Double

    func cost(for order: Order) -> Double {
        base + perKg * order.weightKg
    }
}

/// Context: delegates the pricing decision to the current strategy.
struct Checkout {
    var strategy: ShippingStrategy

    func shipping(for order: Order) -> Double {
        strategy(order)
    }
}

let order = Order(weightKg: 2.4, distanceKm: 120)
let carrier = CarrierRate(base: 3.00, perKg: 2.00)

let strategies: [(String, ShippingStrategy)] = [
    ("flat", flatRate),
    ("weight", weightBased),
    ("distance", distanceBased),
    ("carrier", carrier.cost(for:)),
]

var checkout = Checkout(strategy: flatRate)
for (name, strategy) in strategies {
    checkout.strategy = strategy
    let padded = String(repeating: " ", count: max(0, 8 - name.count)) + name
    print("\(padded): $\(String(format: "%.2f", checkout.shipping(for: order)))")
}
